bundles/org.openhab.binding.cbus/src/main/java/org/openhab/binding/cbus/handler/CBusCGateHandler.java

```diff
@@ -0,0 +1,426 @@
+                ThingHandler handler = thing.getHandler();
+                if (handler instanceof CBusNetworkHandler) {
+                    ((CBusNetworkHandler) handler).cgateStateChanged(isOnline);
+                }
+            }
+        });
+    }
+
+    private void updateGroup(@Nullable String address, @Nullable String value) {
+        if (address == null || value == null) {
+            return;
+        }
+        logger.debug("updateGroup address {}", address);
+        // Address should be of the form //Project/network/application/group
+        if (!address.startsWith("//")) {
+            logger.debug("Address does not start with // so ignoring this update");
+            return;
+        }
+        String[] addressParts = address.substring(2).split("/");
+        if (addressParts.length != 4) {
+            logger.debug("Address is badly formed so ignoring this update length of parts is {} not 4",
+                    addressParts.length);
+            return;
+        }
+        updateGroup(Integer.parseInt(addressParts[1]), Integer.parseInt(addressParts[2]),
+                Integer.parseInt(addressParts[3]), value);
+    }
+
+    private void updateGroup(int network, int application, int group, String value) {
+        for (Thing networkThing : getThing().getThings()) {
+            // Is this networkThing from the network we are looking for...
+            if (networkThing.getThingTypeUID().equals(CBusBindingConstants.BRIDGE_TYPE_NETWORK)) {
+                CBusNetworkHandler netThingHandler = (CBusNetworkHandler) networkThing.getHandler();
+                if (netThingHandler == null || netThingHandler.getNetworkId() != network) {
+                    continue;
+                }
+                // Loop through all the things on this network and see if they match the application / group
+                for (Thing thing : netThingHandler.getThing().getThings()) {
+                    ThingHandler thingThingHandler = thing.getHandler();
+                    if (thingThingHandler == null) {
+                        continue;
+                    }
+
+                    if (thingThingHandler instanceof CBusGroupHandler) {
+                        ((CBusGroupHandler) thingThingHandler).updateGroup(application, group, value);
+                    }
+                }
+            }
+        }
+    }
+
+    public @Nullable CGateSession getCGateSession() {
+        return cGateSession;
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> keepAliveFuture = this.keepAliveFuture;
+        if (keepAliveFuture != null) {
+            keepAliveFuture.cancel(true);
+        }
+        CGateSession cGateSession = this.cGateSession;
+        if (cGateSession != null && cGateSession.isConnected()) {
+            try {
+                cGateSession.close();
+            } catch (CGateException e) {
+                logger.warn("Cannot close CGate session", e);
+            }
+        } else
```

Comment:
please add brackets

Reply:
Fixed this one. I did a search in the code and i cant see any more (but i thought i had done that before).